### ISSUE TITLE
Getting settled, some things that I saw so far

### ISSUE THREAD

**Issue by wzrdtales** [etops/db-migrate-couchbase#1]
# Foreword

Hey there, I'm getting an overview currently and alredy see where things are going in the wrong direction currently. I hope I can clear up your confusion.
# Questions
- Would a docker-compose file not be more suited than a Dockerfile? A Dockerfile is mostly to build the docker container for the current project. I think that could be confusing.
- Which timezone are you living in? Want to know when I can best communicate with you.
# Refs

Refers to https://github.com/db-migrate/node-db-migrate/issues/380

**Comment by wzrdtales** [comment]
Btw. would be awesome if you can give feedback afterwards what you're missing on the docs or even help to make them better. Already giving me feedback what you expect to be delivered to you would help tremendously.

**Comment by wzrdtales** [comment]
Another thing, how do you require your driver currently? It does not seem that you have something in place yet to parse the es6 code?

**Comment by moxious** [comment]
The es6 code works via babel, in the gulp compile target.   It gets translated to es5 in the dist folder, note that package.json refers to the compiled dist/index.js

**Comment by moxious** [comment]
I'm in the EST timezone (GMT-5).  

On Docker -- well actually I may split this out later and reuse a different project I have (https://github.com/moxious/couchbase-docker-circleci) that has an easy docker container, and then just  pull that from circleci and run it.   This code got started before that project was there.

The circleci / docker bits already work great; the tricky part is then how to test. 

Yes, I can provide feedback on the documentation.  I read the entire section on how to contribute, mostly that was pretty good, but of course very skewed towards relational databases.  In the NoSQL part, really what's meant is just mongoDB.  So the docs didn't cover my case but that wasn't really to be expected.   The biggest missing part in the docs for me has to do with the question I already asked -- how to add functions to my driver and use them that don't correspond to the regular MigratorInterface.   

One other bit about docs - one of your recent links showed me where to look, but it's quite difficult for people reading the source if your tests for your drivers are in a different place than the drivers themselves.  I had been assuming you had no unit tests or integration tests in place for the drivers, because I went and looked at the driver code and there was nothing there.  When in fact the tests are still in the main package.

**Comment by wzrdtales** [comment]
yeah you do not have any routines in your project yet I did it actually manually.

actually it throws errors yet

```
[INFO] loading migrations from database
[2016-06-21T18:24:57.571Z]  INFO: db-migrate-couchbase/16610 on tobi: allLoadedMigrations
[2016-06-21T18:24:57.790Z]  INFO: db-migrate-couchbase/16610 on tobi: Close
[ERROR] Error: No primary index on keyspace default. Use CREATE PRIMARY INDEX to create one.
```

I don't know how far you are in development currently, you're probably quite still in setting the project up?

**Comment by moxious** [comment]
This "No primary index" issue has to do with Ottoman, it's easily patched.  I'm actually well past this point because I'm working on other DBs, you just have an empty one.  :)   I will push a fix for that momentarily.

**Comment by wzrdtales** [comment]
> When in fact the tests are still in the main package.

Yes, that is due to the fact, that first with v0.10.x drivers have been moved out, but tests did not yet. That has several reasons, one being that I'm really into doing so much stuff actually (and actually nearly the whole last year though) that I did not yet got into rewriting all tests.

B/c just porting is not enough in this case, vows the used framework has several bugs, which makes it impossible to test some of those things that have been newly introduced. Thus I wanted and needed to switch. Have looked to switch to either mocha or lab. When this switch is being done, this would also include the step moving them away.

**Comment by wzrdtales** [comment]
Hit me up when you pushed something

**Comment by moxious** [comment]
Pushed.   The fix for the issue you had is this:  https://github.com/etops/db-migrate-couchbase/commit/98b2b6ae3b362eae0a67150c7f2a9a617a501d85#diff-168726dbe96b3ce427e7fedce31bb0bcR87

**Comment by moxious** [comment]
BTW -- this mocha testing approach isn't going to work, but this is how I'm attempting to go:  https://github.com/etops/db-migrate-couchbase/blob/master/test/basic_spec.js#L6

**Comment by wzrdtales** [comment]
you should update also the install yourself sh to actually compile the file ;)

**Comment by moxious** [comment]
I just pushed updates to not do docker build at all, to rely on an outside image I control.  No more Dockerfile, init.sh.

Also updated install-yourself to run gulp compile, which was done as a separate step in circle.yml before.

**Comment by wzrdtales** [comment]
Just got this, I get your latest updates currently

```
[2016-06-21T18:42:23.966Z]  INFO: db-migrate-couchbase/20183 on tobi: Connecting to cluster { cluster: 'localhost:8091?detailed_errcodes=1' }
[2016-06-21T18:42:23.967Z]  INFO: db-migrate-couchbase/20183 on tobi: init
[ERROR] TypeError: First argument needs to be a ViewQuery, SpatialQuery or N1qlQuery.
```

**Comment by wzrdtales** [comment]
run-local still relies on couchtest, I doubt you have control about a _ docker image do you?

**Comment by moxious** [comment]
Install yourself is very tricky, sometimes even when I do this, db-migrate won't find the driver.

Just pushed a patch for run-local, I had missed that one.

Whether the docker image is built locally or pulled from somewhere else doesn't really matter, it's just a small sample image with a basic minimum config for testing.

**Comment by wzrdtales** [comment]
I just tried to run your example migration

```
[2016-06-21T18:50:51.293Z]  INFO: db-migrate-couchbase/24023 on tobi: Successfully ran query
[2016-06-21T18:50:51.294Z]  INFO: db-migrate-couchbase/24023 on tobi: 
    runSql { query: 'UPDATE default SET newAttribute=\'FOO\' WHERE _type=\'Test\'',
      params: undefined }
[2016-06-21T18:50:51.305Z]  INFO: db-migrate-couchbase/24023 on tobi: Successfully ran query
[2016-06-21T18:50:51.306Z]  INFO: db-migrate-couchbase/24023 on tobi: 
    renameColumn { ottomanModelName: 'Test',
      oldModelPath: 'renameMe',
      newModelPath: 'renamedColumn' }
[2016-06-21T18:50:51.316Z]  INFO: db-migrate-couchbase/24023 on tobi: Successfully ran query
[2016-06-21T18:50:51.317Z]  INFO: db-migrate-couchbase/24023 on tobi: removeColumn { ottomanModelName: 'Test', modelPath: 'removeMe' }
[2016-06-21T18:50:51.329Z]  INFO: db-migrate-couchbase/24023 on tobi: Successfully ran query
FIRST MIGRATE: UP (done)
[2016-06-21T18:50:51.331Z]  INFO: db-migrate-couchbase/24023 on tobi: Add migration record { name: '/20160618152833-firstMigration' }
[ERROR] 20160618152833-firstMigration { [CouchbaseError: The remote host refused the connection. Is the service up?]
  message: 'The remote host refused the connection. Is the service up?',
  code: 44 }
```

Something is still a bit wrong here though

**Comment by moxious** [comment]
suggests your host config is wrong.  Did you export COUCHBASE_HOST?

**Comment by wzrdtales** [comment]
I actually did not exported anything yet though. But it did ran some queries successfully so I guess it should have a connection

**Comment by moxious** [comment]
Made another tiny push.   Having other config problems here that I'm trying to sort out, related to db-migrate not being able to see the plugin installed.   Here's an example of a successful build on CircleCI where migrations up and down ran successfully:  https://circleci.com/gh/etops/db-migrate-couchbase/53

**Comment by wzrdtales** [comment]
Btw. the most easiest method of developing a driver is actually:
- Do not include db-migrate as a dependency. db-migrate does detect your local version and always prefer the local one over the global installed one.
- use npm link to install your current driver into the global space

I think there are still some troubles about the ES6 stuff though, but I know that there is something to solve this.

**Comment by moxious** [comment]
If you don't install db-migrate as a dependency, then how would you get a copy of yourself via getInstance to do testing later?

**Comment by wzrdtales** [comment]
Yeah I noticed that too just a minute ago. That needs some changes for developer friendliness at db-migrate.

I think the best thing is to have gulp monitoring your files and translating them on the fly. Have yet not really worked with gulp but with grunt though.

**Comment by wzrdtales** [comment]
I think going to the whole getInstance is a bit heavy to do just testing. It is actually doing a whole lot more though.

Looking at what you're currently doing in your driver test you don't really use that much of it although:

So we have this:
https://github.com/etops/db-migrate-couchbase/blob/master/test/basic_spec.js#L15

You don't even need getInstance for this variable. `dataType` and `version` are already exposed to the raw required module.
And the config you could easily pass yourself to the module. So you don't really involve db-migrate itself in the testing process that heavily yet.

**Comment by wzrdtales** [comment]
So at first I look into answering the question for myself:
How far should db-migrate be involved in the testing process of drivers intended for db-migrate. What might be the benefits of having db-migrate itself being involved.

What comes to my mind would be to validate that the Interface stripping actually is properly setup in the driver and works as expected also together with db-migrate. All that side effects might be from interest.

**Comment by wzrdtales** [comment]
So we don't really want db-migrate to be involved in actually calling the drivers functions. But we need some information from db-migrate, like the `dataTypes` also not ultimately necessary though, and we may be want to have some side effects ensured.

From the other side I see the trouble with db-migrate being a dependency of the driver. The best case would be to actually not doing this. To avoid including db-migrate probably dataTypes need to move out of db-migrate. Those could move into the base driver instead, that would be a viable option though. And about side effects, I could mind a generalized set of generic tests that is also used by db-migrate itself published as an additional module.

**Comment by wzrdtales** [comment]
@moxious Do you mind anything else where you currently think you need something of db-migrate to be able to test your driver?

**Comment by wzrdtales** [comment]
Ok for me current tasks are:
- Look to move parts of db-migrate into other modules and try to keep backwards compatibility. 
  That would mean db-migrate-base becoming a direct dependency of db-migrate itself or it might mean an additional module that only serves the dataType. I'm not yet sure about this yet though, that also actually depends on what role dataTypes do play. Actually they do translate into strings, just for this a module may be to much though. I will revisit this and figure out what might be the option that is not introducing some strange relationships.
- Working out the docs afterwards getting clear about testing, delivering examples and a good idea of  how things are done

**Comment by wzrdtales** [comment]
Oh and also give you an example of how your test can be built naturally.

**Comment by moxious** [comment]
I have no opinion on whether DB-migrate should be a dependency for a driver but in order to test drivers correctly it makes sense to do a real migration with them.  That's the testing approach in going in.  Perhaps I don't need unit tests (which would probably require the dependency) if integration tests are covering.

**Comment by moxious** [comment]
After this discussion my two questions are really still the same -- how best to test, and how to add to the default interface provided for SQL databases.  I would think most new drivers would be for nosql DBs since you already have good coverage with popular open source RDBMSs

**Comment by wzrdtales** [comment]
Agreeing to this, but including db-migrate as dependency in a theoretically dependency of this itself is something I would like to avoid.

**Comment by wzrdtales** [comment]
And yes I know that your question are not yet answered, but to give you a proper answer I want to have something in place that really makes sense.

About the Interface stuff, what you have implemented already works. I tested this on your example migration, and it gets everything passed by as expected. I have only one addition to this:

Instead of:

``` javascript
return new Promise((resolve, reject) => {
        return reject('Not implemented');
      }).nodeify(arguments[arguments.length - 1]);
```

you could do also this:

``` javascript
return Promise.reject('Not implemented').nodeify(arguments[arguments.length - 1]);
```

**Comment by wzrdtales** [comment]
To note:

You have added also `runN1ql`, but you do not have such a function you do have a `_runN1ql`, so this listed function just gets your defined dummy passed.

But for example `registerModel` got passed as expected. I logged from the migration the following input which was passed, which is not the dummy just as expected:

``` javascript
function registerModel(modelName, ottomanModel) {
    this.models[modelName] = ottomanModel;
    return this;
  }
```

**Comment by moxious** [comment]
OK thanks.  I understand what you mean on the potential circular dependency, and if there's a better pattern that lets me test a driver in a realistic context I'll readily adopt it.

I pushed small fixes for the promise structure and the _runN1ql issue you identified.  I'll keep experimenting tonight and tomorrow to see what I can find, and wait to hear back from you on an appropriate pattern for testing.

**Comment by wzrdtales** [comment]
I will come back to you later on again, its quite late over here (23:22) and I want to finish some other things in the next hours before I go sleep. I will come back on this topic may be already with a solution tomorrow.

**Comment by wzrdtales** [comment]
So ok, to come back to you.

I have made my decision already, I will move some parts into a new package called `db-migrate-shared`. This will include currently the default log interface and dataTypes, but might get as well extended.

So you can really concentrate on testing your driver and don't should need to care that much about db-migrate. So here is how I will come up with the issue to test if db-migrate would operate as expected with your driver. I will provide a default test suite, that tests only a few things that is expected from any driver.
This is for instance the connect method, which should result in an usable driver, The existence of the close method and expecting to return the Promise or call the callback properly.

Up from this everything should be straight forward just testing every single function of your driver and after that checking against the db if it has the expected result. So quite similar to all current driver tests like the ones in the db-migrate repo:
https://github.com/db-migrate/node-db-migrate/blob/master/test/driver/mongodb_test.js

**Comment by moxious** [comment]
OK, thanks.  I will continue on further testing, and see how far I can get.

**Comment by wzrdtales** [comment]
I have just quickly refactored those pieces, you can also already include the `db-migrate-shared` module to get the dataTypes from it.

**Comment by moxious** [comment]
Will get to this tomorrow.  I'm working on pushing changes to improve unit testing right now, things are getting better.  https://circleci.com/gh/etops/db-migrate-couchbase/65

**Comment by moxious** [comment]
@wzrdtales If you have time, please take a look.  I've made quite a lot of changes since the above.  I rebased git the squash a lot of this stuff so have a fresh pull.  Unit tests suite passes on circleCI on a live database, with migrations running up and down.   Decent test coverage (80%) can be seen in the artifacts of this build:  https://circleci.com/gh/etops/db-migrate-couchbase/79#artifacts

For what I set out to do, primarily I think I need to do more testing, more migrations, more scenarios but the basic structure is there and the code seems to be working.  What else would you recommend?

**Comment by moxious** [comment]
In particular, I'm wondering about how for a non-SQL database, options like dry-run should be handled, and what the `--sql-file` option should do with a driver like this, or how the two should interact.   Perhaps too I'm still missing some stuff on the seeds table, or how seeds would work.

**Comment by wzrdtales** [comment]
Well sql-files would not apply to your driver unless the db supports the sql query language. And dry-run should be handled like it would be for a rdbms.

On a dry run you only do certain things. For example, create the migrations table, and get the records from that migrations table. Everything else just does nothing and returns the query. As for you there is nothing like an query, you could output informations that describe in either a common dialogue of the database what is going to happen, or just describe the changes going to be made. Basically everything you add for dry-run, you could also add to be displayed on verbose runs.

**Comment by wzrdtales** [comment]
So `--sql-file` should in your case output a warning to the user in the console like:

```
This driver does not support sql files.
```

**Comment by wzrdtales** [comment]
Ok, forget what I've said. You can actually completely ignore `--sql-file`, this is never going to reach your driver. It is only handled in db-migrate and is only related to the create command of db-migrate. Just thought into the wrong direction for a second.

**Comment by wzrdtales** [comment]
Maybe this helps to make clear what I expect from a No/NewSQL driver.

A NoSQL driver does at least implement the default API with the nearest equivalent available. In the case of mongodb `createCollection` comes close to createTable and thus translates to this. This expectation is made to give a very basic connection between different database. No matter what type they actually are.

But for most of the time, one chooses a database for a reason and having just some generic interfaces wont help him to use the full potential of the database. And given this reason, this compatibility between different databases should be just limited, but given. So you don't need to support everything, but everything you can.

**Comment by moxious** [comment]
> A NoSQL driver does at least implement the default API with the nearest equivalent available. In the case of mongodb createCollection comes close to createTable and thus translates to this. This expectation is made to give a very basic connection between different database. No matter what type they actually are.

This part I think I have.  The mapping between the ideas isn't perfect, but it seems reasonable and it's documented in the driver's README.

I think I'm almost to the point where I've supported everything I can from the db-migrate interface.  There is extra stuff I would choose to support beyond that interface, but that's OK we have discussed that.  I mostly now need to build confidence the the tests are suitable and appropriate so that I might use this on a real database without fear of creating bigger problems than the modules solves.

**Comment by moxious** [comment]
@wzrdtales initial version published; https://www.npmjs.com/package/db-migrate-couchbase

**Comment by wzrdtales** [comment]
@moxious Cool thing :)

One thing left, you should look to remove this dependency: https://github.com/etops/db-migrate-couchbase/blob/master/package.json#L28

and replace it with the new module named above https://www.npmjs.com/package/db-migrate-shared

This dependency seems to be unused right now anyway. But what you were originally getting from it now resides in `db-migrate-shared`

**Comment by wzrdtales** [comment]
@moxious Cleaning up the issues currently. Closing here, but feel free to hit me up anytime you've a question.

**Comment by wzrdtales** [comment]
oh wrong repo

**Comment by moxious** [comment]
:)  No problem will do
